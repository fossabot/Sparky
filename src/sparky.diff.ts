export function populateDOMwithDiff(currentDom: HTMLElement, dom: HTMLElement) {
    if(!currentDom) return dom;

    const currentNodesStack = [{node: currentDom, parent: null}];
    const nextDomStack = [{node: dom, parent: null}];

    while(currentNodesStack.length > 0 || nextDomStack.length > 0) {
        let currentStack = currentNodesStack.pop();
        let nextStack = nextDomStack.pop();

        let node = currentStack.node;
        let nextNode = nextStack.node;

        if(node && !nextNode) {
            node.remove();
            return;
        }

        if(nextNode && !node) {
            (currentStack.parent as HTMLElement).appendChild(nextNode);
            return;
        }

        if(node.nodeName != nextNode.nodeName) {
            node = diffName(nextNode, node);
        }

        if(node.nodeName !== "#text") {
            diffAttribute(node, nextNode);            
        }

        if(node.children.length > 0) {
            Array.from(node.children).map((child: HTMLElement) => currentNodesStack.push({node: child, parent: node}))
        }

        if(nextNode.children.length > 0) {
            Array.from(nextNode.children).map((child: HTMLElement) => nextDomStack.push({node: child, parent: nextNode}))
        }
    }
}

function diffName(nextNode: HTMLElement, node: HTMLElement) {
    const newNode = document.createElement(nextNode.nodeName);
    Array.from(node.attributes).map((attr: Attr) => {
        newNode.attributes.setNamedItem(attr.cloneNode(true) as Attr);
    });
    newNode.innerHTML = node.innerHTML;
    node = newNode;
    return node;
}

function diffAttribute(node: HTMLElement, nextNode: HTMLElement) {
    Array.from(node.attributes).forEach((attr: Attr, i: number) => {
        const nextAttr = nextNode.attributes[i];
        if (attr.name != nextAttr.name) {
            node.attributes.removeNamedItem(attr.name);
            node.attributes.setNamedItem(nextAttr.cloneNode(true) as Attr);
        }
        if (attr.value != nextAttr.value) {
            attr.value = nextAttr.value;
        }
    });
}
