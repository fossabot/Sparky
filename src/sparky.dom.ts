export function reconciliate(currentDom: HTMLElement, newDom: HTMLElement) {
    if (!currentDom && !newDom) return null;
    if (currentDom && !newDom) return null;
    if (!currentDom && newDom) return newDom;

    const currentDomStack = [currentDom];
    let newDomStack = [newDom];

    let currentDomToAttach = currentDom;

    while (currentDomStack.length > 0 || newDomStack.length > 0) {
        let dom = currentDomStack.shift();
        let nextDom = newDomStack.shift();

        if (dom && !nextDom) {
            dom.parentElement.removeChild(dom);
            continue;
        }

        if (!dom && nextDom) {
            currentDomToAttach.appendChild(nextDom);
            continue;
        }

        if(dom.nodeName == "#text" && nextDom.nodeName != "#text") {
            currentDomToAttach.removeChild(dom);
            currentDomToAttach.appendChild(nextDom);
            continue;
        }
        
        setAttributes(dom, nextDom);

        if (dom.nodeName !== nextDom.nodeName) {
            dom.parentElement.replaceChild(nextDom, dom);
        }

        if ((dom.nodeName == "#text" && nextDom.nodeName == "#text") 
        || (nextDom.childNodes.length == 1 && nextDom.childNodes[0].nodeName == "#text")) {
            if (dom.textContent !== nextDom.textContent)
                dom.textContent = nextDom.textContent
        }

        if (dom.childNodes.length > 0) {
            currentDomToAttach = dom;
            Array.from(dom.childNodes).forEach((child: HTMLElement) => {
                if(dom.childNodes.length == 1 && dom.childNodes[0].nodeName == "#text") return;
                currentDomStack.push(child)
            })
        }

        if (nextDom.childNodes.length > 0) {
            if (dom)
                currentDomToAttach = dom;
            Array.from(nextDom.childNodes).forEach((child: HTMLElement) => {
                if(dom.childNodes.length == 1 && dom.childNodes[0].nodeName == "#text") return;
                newDomStack.push(child)
            })
        }
    }

    return currentDom;
}

function setAttributes(currentDom: HTMLElement, nextDom: HTMLElement) {
    const currentAttributesList = currentDom && currentDom.attributes
        ? transformAttributesToSortedArray(currentDom.attributes) : [];

    const nextAttributesList = nextDom && nextDom.attributes
        ? transformAttributesToSortedArray(nextDom.attributes) : [];

    while(nextAttributesList.length > 0) {
        const currentAttr = currentAttributesList.shift();
        const nextAttr = nextAttributesList.shift();

        if(currentAttr && !nextAttr) {
            currentDom.removeAttribute(currentAttr.name);
            continue;
        }

        if(!currentAttr && nextAttr) {
            currentDom.setAttribute(nextAttr.name, nextAttr.value);
            continue;
        }

        if(currentAttr.name != nextAttr.name) {
            currentDom.removeAttribute(currentAttr.name);
            currentDom.setAttribute(nextAttr.name, nextAttr.value);
            continue;
        }

        if(currentAttr.value != nextAttr.value) {
            currentDom.getAttributeNode(nextAttr.name).value = nextAttr.value;
        }
    }
    
    if(currentAttributesList.length > 0) {
        currentAttributesList.forEach(attr => currentDom.attributes.removeNamedItem(attr.name))
    }
}

function transformAttributesToSortedArray(arrayLike: ArrayLike<Attr>) {
    return Array.from(arrayLike).sort((a: Attr, b: Attr) => {
        return a.name.localeCompare(b.name);
    })
}