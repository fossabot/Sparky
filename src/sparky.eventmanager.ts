interface eventListSingle {
    dom: HTMLElement;
    type: string;
    callbackFn: eventCallbackFn;
}

type eventCallbackFn = (evt: Event) => void;


export class EventManager {
    static oldEventType: string[] = [];
    static eventList: eventListSingle[] = [];
    static eventListType: string[] = [];

    static listen() {
        EventManager.removeUnusedEvents();
        this.eventList.forEach(({type}) => {
            if(!this.isEventTypeListening(type)) {
                document.body.addEventListener(type, (event) => this.dispatchEvent(event))
                this.eventListType.push(type);
            }
        });
    }

    static dispatchEvent(event: Event) {
        this.eventList.find((evtList) => {
            if(this.isEventTarget(evtList, event)) {
                if(evtList.type === event.type) {
                    evtList.callbackFn(event);
                }
            }
        })
    }

    static addEvent(eventSingle: eventListSingle) {
        const {dom, type, callbackFn} = eventSingle;
        this.eventList.push({ dom, type, callbackFn });
    }

    private static removeUnusedEvents() {
        this.eventList = this.eventList
            .filter((evt) => evt.dom.isConnected);
    }

    private static isEventTypeListening(type: string) {
        return this.eventListType.find(t => t == type);
    }

    private static isEventTarget(evtList: eventListSingle, event: Event) {
        return (evtList.dom === event.target) || evtList.dom.contains(event.target as HTMLElement);
    }
}