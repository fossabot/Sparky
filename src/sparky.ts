import { SparkyFunction } from "./sparky.function";
import { diff } from "./sparky.dom";
import { setAllEvents } from "./sparky.event";
import { EventManager } from "./sparky.eventmanager";

export interface IRenderReturn {
    dom: HTMLElement,
    func: Function[]
}

export type ISelfFunction = (self: SparkyFunction) => IRenderReturn;

export interface ISparkyComponent {
    self: SparkyFunction;
    selfFn: ISelfFunction;
}

export class Sparky {
    private static currentDom: HTMLElement;

    /**
     * Generate a Sparky Component that can be mount.
     * @param renderFunc The function that going to be execute to render html template
     */
    static component(renderFunc: ISelfFunction) {
        const thisFunction = new SparkyFunction(renderFunc);
        return { self: thisFunction, selfFn: renderFunc } as ISparkyComponent;
    }

    /**
     * Mount a Sparky Component in the DOM Tree and keep it hydrate.
     * @param component Sparky Component
     * @param dom The dom element where you want to mount this component
     */
    static mount(component: ISparkyComponent, dom?: HTMLElement) {
        const { self, selfFn } = component;
        const render = selfFn.call(window, self) as IRenderReturn;

        let finalDOM = diff(this.currentDom, render.dom);
        finalDOM = setAllEvents({dom: finalDOM, func: render.func}, self);
        if (!finalDOM) return;
        EventManager.listen();
        if (!finalDOM.isConnected && dom)
            dom.appendChild(finalDOM);
        this.currentDom = finalDOM as HTMLElement;
    }

    static diff(oldNode: HTMLElement, newNode: HTMLElement) {
        return diff(oldNode, newNode)
    }
}

/**
 * Render the html string template to HTML elements
 * @param html Array of HTML String 
 * @param computedProps Computed Props used to pass Javascript into template
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals
 */
export function render(html: TemplateStringsArray | string, ...computedProps: any[]): IRenderReturn {
    const domNode = document.createElement("div");
    const func: Function[] = [];

    const newHTML = (typeof html == "string") ? html
        : html.map((stringHTML, i) => {
            let htmlLine = ""
            htmlLine += stringHTML
            if(!computedProps[i]) return htmlLine;
            if (typeof computedProps[i] == "function") {
                func.push(computedProps[i])
                htmlLine += "'functionScoped'";
            } else if(computedProps[i].type && computedProps[i].type == "SparkyComponent") {
                const comp = computedProps[i] as ISparkyComponent;
                const render =  comp.selfFn.call(window, comp.self) as IRenderReturn;
                
                const div = document.createElement("div");
                div.appendChild(render.dom)

                func.push(...render.func);
                htmlLine += div.innerHTML;
            } else {
                computedProps[i] = new String(computedProps[i]);
                if ((computedProps[i] as string).startsWith("<"))
                    htmlLine += computedProps[i]
                else
                    htmlLine += `<span class='computed'>${computedProps[i]}</span>`
            }
            return htmlLine;
        })
        
    domNode.innerHTML = Array.isArray(newHTML) ? newHTML.join("") : newHTML;
    return { dom: domNode.firstElementChild as HTMLElement, func };
}