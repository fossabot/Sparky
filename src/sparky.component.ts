import { IRenderReturn, ISparkyComponent, renderToDOMNode, IStateRoute } from "./sparky";
import { findEvent } from "./sparky.event";
import { eventCallbackFn, IEventSingle } from "./sparky.eventmanager";
import { SparkyContext } from "./sparky.context";
import { IRoutingTypes } from "./sparky.router";

export interface IParams {
    [x: string]: string
}
export interface ISparkyRoot {
    id: string;
    type: IRoutingTypes;
    basename: string;
    forceURLUpdate: boolean;
    historyIndex: number;
    stateChanging: boolean;
    params: IParams[];
    history: IStateRoute[];
    routing: IStateRoute[];
    updateAt: number;
}

export interface HTMLElementSparkyEnhanced extends HTMLElement {
    __sparkyEvent?: IEventSingle;
    __sparkyRoot?: ISparkyRoot;
}

interface ICachedComponent {
    component: ISparkyComponent,
    dom: HTMLElement
}

export class SparkyComponent {
    private static cachedComponent: ICachedComponent[][] = [];
    static populate(nextDOM: HTMLElement, render: IRenderReturn, rootComponent: ISparkyComponent) {
        
        const renderQueue: [IRenderReturn, ISparkyComponent, HTMLElement][] = [[render, rootComponent, nextDOM]];

        let depthHorizontal = 0;

        while (renderQueue.length > 0) {
            const [currentRender, currentComponent, currentDOM] = renderQueue.shift();
            if(!this.cachedComponent[depthHorizontal]) this.cachedComponent[depthHorizontal] = [];

            currentRender.func.forEach((currentFunc, index) => {
                const currentEvent = findEvent(currentDOM, currentRender.renderId, index);
                const eventName = currentEvent.attr.name.replace("on", "");
                currentEvent.dom.__sparkyEvent = {
                    type: eventName,
                    context: currentComponent.context,
                    callbackFn: currentFunc.func as eventCallbackFn
                };
                currentEvent.dom.removeAttribute(currentEvent.attr.name)
            })

            currentRender.children.forEach((currentChild) => {
                renderQueue.push([currentChild, currentComponent, currentDOM])
            })

            currentRender.nestedComponents.forEach((currentComp, index) => {
                const cached = this.cachedComponent[depthHorizontal][index];
                const commentDom = this.findComment(currentDOM, currentRender.renderId, index, currentComp.renderFn.name);
                if(cached) {
                    if(cached.component.renderFn.name == "") {
                        if(cached.component.renderFn.toString() == currentComp.renderFn.toString()) {
                            currentComp = getCachedComponent(cached, currentComp);
                        }
                    } else if(cached.component.renderFn.name == currentComp.renderFn.name){
                        currentComp = getCachedComponent(cached, currentComp);
                    }
                }
                SparkyContext.setCurrentContext(currentComp.context);
                SparkyContext.resetIndexes();
                currentComp.context.__root = rootComponent;
                currentComp.context.__rootElement = rootComponent.context.__rootElement;
                const renderChild = currentComp.renderFn(Object.freeze(currentComp.context.props));
                if(!commentDom) return;
                const renderChildDOM = renderToDOMNode(renderChild.html);
                commentDom.parentNode.replaceChild(renderChildDOM, commentDom);
                render.func.push(...renderChild.func);
                renderQueue.push([renderChild, currentComp, renderChildDOM]);

                this.cachedComponent[depthHorizontal][index] = {
                    component: currentComp,
                    dom: currentDOM
                };
            });
            if(currentRender.nestedComponents.length > 0)
                depthHorizontal++;
        }
        
        return nextDOM;
    }

    private static findComment(element: HTMLElement, renderId: string, i: number, componentName: string) {
        const domQueue = [element];
        while(domQueue.length > 0) {
            const elem = domQueue.shift();
            if(elem.nodeName == '#comment' && elem.nodeValue.trim() == `SparkyComponent#${componentName}#${i}#${renderId}`) {
                return elem;
            }

            for(let i = 0; i < elem.childNodes.length; i++) {
                domQueue.push(elem.childNodes[i] as HTMLElement);
            }
        }
    }
}

function getCachedComponent(cached: ICachedComponent, currentComp: ISparkyComponent) {
    cached.component.context.props = currentComp.context.props;
    currentComp = cached.component;
    return currentComp;
}
