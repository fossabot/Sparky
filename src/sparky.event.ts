import { SparkyFunction } from "./sparky.function";
import { IRenderReturn } from "./sparky";

export function setEvents(render: IRenderReturn, self: SparkyFunction) : HTMLElement {
    const { dom, func } = render;
    Array.from(dom.attributes).forEach((attr: Attr) => {
        if(attr.name.startsWith("on")) {
            const execFun = func.shift();
            dom.addEventListener(attr.name.replace("on", ""), (event) => {    
                execFun.call(self, event)
            })
            dom.attributes.removeNamedItem(attr.name)
        }
    })
    return dom;
}

export function setAllEvents(render: IRenderReturn, self: SparkyFunction) : HTMLElement {
    const { dom, func } = render;
    const currentStack = [dom];
    while(currentStack.length > 0) {
        const ele = currentStack.pop();
        setEvents({dom: ele, func: render.func}, self);
        if(ele.children.length > 0) {
            Array.from(ele.children).map((child: HTMLElement) => currentStack.push(child))
        }
    }
    return dom;
}