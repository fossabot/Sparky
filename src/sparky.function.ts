import { Sparky, IRenderReturn } from "./sparky";

type UpdateCallback = () => void;

type DependenciesList = string[];

export class SparkyFunction {
    private newProps: string[] = [];
    private state: any;
    private renderFunc: (self: SparkyFunction) => IRenderReturn;

    constructor(renderFunc: (self: SparkyFunction) => IRenderReturn) {
        this.state = {};
        this.renderFunc = renderFunc;
    }

    /**
     * Execute after the render/update of the DOM tree.
     * @param callback - The function that you want to execute
     * @param dependenciesChanged - An array of keys to know when the onUpdate need to be executed
     */
    onUpdate = (callback: UpdateCallback, dependenciesChanged?: DependenciesList) => {
        if (!dependenciesChanged && this.newProps.length == 0 ||
            dependenciesChanged && dependenciesChanged.length == 0 ||
            this.newProps.some((props) => dependenciesChanged && dependenciesChanged.includes(props)))
            window.requestIdleCallback(() => callback.call(this))
    }

    /**
    * Get State object value of this context
    * @param props - the specific key of the value that you want to retrieve
    */
    getState = <S>(props: string): S => {
        if (props) return this.state[props];
        return this.state;
    }

    /**
     * Add/Set a new value into the State object of the context
     * @param newState - new Value
     */
    setState = <S>(newState: S) => {
        this.newProps = Object.keys(newState);
        this.state = { ...this.state, ...newState };
        Sparky.mount({
            self: this,
            selfFunc: this.renderFunc
        });
    }
}