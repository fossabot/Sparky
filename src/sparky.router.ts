import { IStateRoute, Sparky } from "./sparky"
import { HTMLElementSparkyEnhanced, ISparkyRoot } from "./sparky.component";

declare var thisTest;
export type IRoutingTypes = "browser" | "hash" | "abstract";

export interface documentSparkyEnhanced extends Document {
    __sparkyRoutingId: string;
}

export function listeningHashChange(stateRoute: IStateRoute[], callbackFn: Function, dom: HTMLElementSparkyEnhanced) {
    window.requestIdleCallback(() => {
        window.addEventListener("hashchange", (evt) => {
            changeStateByEvent(evt, stateRoute, callbackFn, dom)
        })
    })
    if(typeof thisTest != "undefined" && thisTest.testing) {
        const evt = new Event("hashchange") as HashChangeEvent;
        Object.defineProperty(evt, 'oldUrl', {writable: false, value: ""});
        Object.defineProperty(evt, 'newURL', {writable: true, value: "#"});
        changeStateByEvent(evt, stateRoute, callbackFn, dom);
        dom.__sparkyRoot.stateChanging = true;
        dom.__sparkyRoot.forceURLUpdate = true;
        changeStateByEvent(evt, stateRoute, callbackFn, dom)
    }
}

function changeStateByEvent(evt: HashChangeEvent, stateRoute: IStateRoute[], callbackFn: Function, dom: HTMLElementSparkyEnhanced) {
    const documentSparky = document as documentSparkyEnhanced;

    if(dom.__sparkyRoot.forceURLUpdate) 
        documentSparky.__sparkyRoutingId = dom.__sparkyRoot.id;

    if(documentSparky.__sparkyRoutingId 
        && documentSparky.__sparkyRoutingId != dom.__sparkyRoot.id) {
            console.warn("Only one route object can have the control of URL")
            return;
        }
    if(evt.oldURL == evt.newURL || dom.__sparkyRoot.stateChanging) {
        dom.__sparkyRoot.stateChanging = false;
        return;
    };
    const newState = getStateByHash(stateRoute, location.hash);
    newState.hash = location.hash;
    pushToAbstractHistory(dom.__sparkyRoot, newState);
    if(newState) {
        callbackFn(newState.component);
    }
    documentSparky.__sparkyRoutingId = dom.__sparkyRoot.id;
}

export function getStateByHash(stateRoute: IStateRoute[], newPath: string) {
    return stateRoute.find((state, i) => {
        if (state.path instanceof RegExp) {
            return newPath.search(state.path) != -1;
        }
        else if (typeof state.path == "string") {
            return newPath.includes(state.path);
        }
        else {
            return false;
        }
    }) || stateRoute[0];
}


export function pushToAbstractHistory(sparkyRoot: ISparkyRoot, stateRoute: IStateRoute) {
    if(sparkyRoot.historyIndex < (history.length - 1)) {
        sparkyRoot.history = sparkyRoot.history.slice(0, sparkyRoot.historyIndex + 1)
    }
    sparkyRoot.history.push(stateRoute);
    sparkyRoot.historyIndex = sparkyRoot.history.length - 1;
}

export function Sparky_cleanHistory(this: HTMLElementSparkyEnhanced) {
    this.__sparkyRoot.history = [];
    this.__sparkyRoot.historyIndex = 0;
    this.__sparkyRoot.stateChanging = true;
    location.hash = "";
} 

export function Sparky__goToState(this: HTMLElementSparkyEnhanced, newPath: string) {
    let { routing, type } = this.__sparkyRoot;
    const routeState = getStateByHash(routing, newPath);
    routeState.hash = newPath;
    this.__sparkyRoot.stateChanging = true;
    let normalizePath = newPath;
    switch(type) {
        case "hash": location.hash = "/#/" + normalizePath; break;
        case "browser": location.hash = normalizePath; break;
    }
    pushToAbstractHistory(this.__sparkyRoot, routeState);
    Sparky.mount(routeState.component, this);
}

export function Sparky__back(this: HTMLElementSparkyEnhanced) {
    let { history, type } = this.__sparkyRoot;
    if (this.__sparkyRoot.historyIndex - 1 < 0) return;
    const state = history[--this.__sparkyRoot.historyIndex];
    this.__sparkyRoot.stateChanging = true;
    let normalizePath = state.hash;
    switch(type) {
        case "hash": location.hash = "/#/" + normalizePath; break;
        case "browser": location.hash = normalizePath; break;
    }
    Sparky.mount(state.component, this);
}

export function Sparky__forward(this: HTMLElementSparkyEnhanced) {
    let { history, type } = this.__sparkyRoot;
    if (this.__sparkyRoot.historyIndex + 1 > history.length - 1) return; 
    const state = history[++this.__sparkyRoot.historyIndex];
    this.__sparkyRoot.stateChanging = true;
    let normalizePath = state.hash;
    switch(type) {
        case "hash": location.hash = "/#/" + normalizePath; break;
        case "browser": location.hash = normalizePath; break;
    }
    Sparky.mount(state.component, this)
}