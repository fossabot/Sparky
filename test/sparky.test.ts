import { renderToDOMNode, state, memoize, update, render, Sparky } from "../src/sparky";
import { SparkyTest } from "../src/sparky.test-util";
import { EventManager } from "../src/sparky.eventmanager";

const lib = require("../src/sparky");

describe("Diff method", () => {
    test("Nothing to render", () => {
        const diff = Sparky.reconciliate(null, null);
        expect(diff).toBe(null);
    });
    test("Only oldNode provided", () => {
        const diff = Sparky.reconciliate(renderToDOMNode(render(`<div>Hola a todos</div>`).html), null);
        expect(diff).toBe(null);
    });

    test("Only newNode provided", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos</div>`).html);
        const diff = Sparky.reconciliate(null, currentDom);
        expect(diff).toBe(currentDom);

        expect(diff.outerHTML).toBe(render(`<div>Hola a todos</div>`).html);
    });

    test("If same nodes are provided the result must be the same", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos</div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos</div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos</div>`).html);
    });

    test("Add a new node by diff", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos</div>`).html);
        const diff = Sparky.reconciliate(currentDom,
            renderToDOMNode(render(`<div>Hola a todos <span>Jeremy</span></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <span>Jeremy</span></div>`).html);
    });

    test("Remove a node by diff", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <span>Pablo</span><span>Jeremy</span></div>`).html);
        const diff = Sparky.reconciliate(currentDom,
            renderToDOMNode(render(`<div>Hola a todos <span>Jeremy</span></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <span>Jeremy</span></div>`).html);
    });

    test("Replace a existing node by diff", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <b>Before</b></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <em>After</em></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <em>After</em></div>`).html);
    });

    test("Adding a new nodes as a list by diff", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <ul></ul></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <ul>
            <li>Hola</li>
            <li>Adios</li>
            <li>Buenas</li>
        </ul></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <ul>
            <li>Hola</li>
            <li>Adios</li>
            <li>Buenas</li>
        </ul></div>`).html);
    });

    test("Add new attributes to an element", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <b>Before</b></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <b class='selected'>After</b></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <b class=\"selected\">After</b></div>`).html);
    });

    test("Add more attributes to an element", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <b contenteditable='true' class='selected'>Before</b></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <b id='uno' class='selected more'>After</b></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <b class=\"selected more\" id=\"uno\">After</b></div>`).html);
    });

    test("Add less attributes to an element", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <b id='uno' class='selected more'>Before</b></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <b class='selected'>After</b></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <b class=\"selected\">After</b></div>`).html);
    });

    test("Change attributes to an element", () => {
        const currentDom = renderToDOMNode(render(`<div>Hola a todos <b class='selected more'>Before</b></div>`).html);
        const diff = Sparky.reconciliate(currentDom, renderToDOMNode(render(`<div>Hola a todos <b class='selected'>After</b></div>`).html));

        expect(diff).toBe(currentDom);
        expect(diff.outerHTML).toBe(render(`<div>Hola a todos <b class=\"selected\">After</b></div>`).html);
    });
})

describe("Mount function", () => {
    test("Basic function", () => {
        Sparky.mount(Sparky.component(() => {
            return lib.render /*html*/`
            <div>
                <p>Hola a todos</p>
            </div>
            `
        }), document.body);
        expect(document.body).toMatchSnapshot()
    });
    test("Computed function", () => {
        Sparky.mount(Sparky.component(() => {
            const a = 24;
            return render /*html*/`
            <div>
                <p>Hola a todos ${a}</p>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed w/ State function", () => {
        Sparky.mount(Sparky.component(() => {
            const [a, setA] = state("uno");
            return render /*html*/`
            <div>
                <p>Hola a todos ${a}</p>
            </div>
            `
        }), document.body);
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed, State w/ Event function", () => {
        const test = SparkyTest.test(() => {
            Sparky.mount(Sparky.component(() => {
                const [a, setA] = state("uno");
                const onClick = () => {
                    setA("10");
                }
                return render /*html*/`
                <div>
                    <p id="unique" onclick=${onClick}>Hola a todos ${a}</p>
                </div>
                `
            }), document.body);
        });
        test.selector("#unique").simulate("click");
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
    test("Computed, State, Memo w/ Event and Click function", () => {

        const Span = () => {
            const [a, setA] = state("uno");

            const onClick = () => {
                setA("100")
            }

            return render /*html*/`
            <div>
                <p id="paragraph-1" onclick=${onClick}>Hola a todos ${a}</p>
            </div>
            `
        };

        const test = SparkyTest.test(() => {
            Sparky.mount(Sparky.component(() => {
                const [a, setA] = state("hola mundo");

                memoize(() => {
                    console.log("Memo!")
                }, [a]);

                update(() => {
                    console.log("Update!")
                })

                const onClick = () => {
                    setA("10")
                }

                return render /*html*/`
            <div>
                <p id="paragraph" onclick=${onClick}>Hola a todos ${a}</p>
                <div>${Sparky.component(Span)}</div>
            </div>
            `
            }), document.body);
        });

        test.selector("#paragraph").simulate("click");
        test.selector("#paragraph-1").simulate("click");
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });

    test("Computed, State w/ Event function with Render inside", () => {
        const test = SparkyTest.test(() => {
            Sparky.mount(Sparky.component(() => {
                const [a, setA] = state("uno");
                const onClick = () => {
                    setA("10");
                }
                update(() => {
                    console.log("Update!")
                }, [a, "Jeremy"])
                return render /*html*/`
                <div>
                    <p id="unique" onclick=${onClick}>Hola a todos ${a}</p>
                    <div>${a == "10" ? render `<div>Hola</div>` : "No hay nada"}</div>
                    <div>${Sparky.component(() => {
                        return render `<div>Esto es un componente</div>`;
                    })}</div>
                </div>
                `
            }), document.body);
        });
        test.selector("#unique").simulate("click");
        // console.log(document.body.innerHTML)
        expect(document.body).toMatchSnapshot()
    });
})