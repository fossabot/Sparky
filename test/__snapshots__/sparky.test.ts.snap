// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount function Basic function 1`] = `
<div
  id="app"
>
  <div>
    
                
    <p>
      Hola a todos
    </p>
    
            
  </div>
</div>
`;

exports[`Mount function Computed function 1`] = `
<div
  id="app"
>
  <div>
    
                
    <p>
      Hola a todos 
      <span
        class="computed"
      >
        24
      </span>
    </p>
    
            
  </div>
</div>
`;

exports[`Mount function Computed w/ State function 1`] = `
<div
  id="app"
>
  <div>
    
                
    <p>
      Hola a todos 
      <span
        class="computed"
      >
        uno
      </span>
    </p>
    
            
  </div>
</div>
`;

exports[`Mount function Computed, State w/ Event function 1`] = `
<div
  id="app"
>
  <div>
    
                    
    <p
      id="unique"
    >
      Hola a todos 
      <span
        class="computed"
      >
        10
      </span>
    </p>
    
                
  </div>
</div>
`;

exports[`Mount function Computed, State w/ Event function with Render inside 1`] = `
<div
  id="app"
>
  <div>
    
                    
    <p
      id="unique"
    >
      Hola a todos 
      <span
        class="computed"
      >
        10
      </span>
    </p>
    
                    
    <div>
      <div>
        Hola
      </div>
    </div>
    
                    
    <div>
      <div>
        Esto es un componente
      </div>
    </div>
    
                
  </div>
</div>
`;

exports[`Mount function Computed, State, Memo w/ Event and Click function 1`] = `
<div
  id="app"
>
  <div>
    
                
    <p
      id="paragraph"
    >
      Hola a todos 
      <span
        class="computed"
      >
        10
      </span>
    </p>
    
                
    <div>
      <div>
        
                
        <p
          id="paragraph-1"
        >
          Hola a todos 
          <span
            class="computed"
          >
            100
          </span>
        </p>
        
            
      </div>
    </div>
    
            
  </div>
</div>
`;

exports[`Mount function Routing Functionality with Browser History 1`] = `
<div
  id="app3"
>
  <div>
     
    <span
      id="unique1"
    >
      GoTo
    </span>
     
    <span
      id="unique-back"
    >
      Everyone
    </span>
  </div>
</div>
`;

exports[`Mount function Routing Functionality with Hash History 1`] = `
<div
  id="app2"
>
  <div>
     
    <span
      id="unique1"
    >
      GoTo
    </span>
     
    <span
      id="unique-back"
    >
      Everyone
    </span>
  </div>
</div>
`;
